src/gfx/gizmos/rotation_gizmo.hpp

```diff
@@ -0,0 +1,184 @@
+
+      selectionDisc.center = extractTranslation(transform);
+    }
+  }
+
+  size_t getHandleIndex(Handle &inHandle) { return size_t(inHandle.userData); }
+
+  float3 getAxisDirection(size_t index, float4x4 transform) {
+    float3 base{};
+    base[index] = 1.0f;
+    return linalg::normalize(linalg::mul(transform, float4(base, 0)).xyz());
+  }
+
+  // Called when handle grabbed from IGizmoCallbacks
+  virtual void grabbed(InputContext &context, Handle &handle) {
+    dragStartTransform = transform;
+
+    size_t index = getHandleIndex(handle);
+    SPDLOG_DEBUG("Handle {} ({}) grabbed", index, getAxisDirection(index, dragStartTransform));
+
+    auto &selectionDisc = handleSelectionDiscs[index];
+    float d;
+    // Check if ray intersects with plane of the disc (either side)
+    if (intersectPlane(context.eyeLocation, context.rayDirection, selectionDisc.center, selectionDisc.normal, d) ||
+        intersectPlane(context.eyeLocation, context.rayDirection, selectionDisc.center, -selectionDisc.normal, d)) {
+      float3 hitPoint = context.eyeLocation + d * context.rayDirection;
+      float3 radiusVec = hitPoint - selectionDisc.center;
+
+      // Construct a plane whose normal is the vector from hitPoint to context.eyeLocation and project tangent onto this plane
+      dragStartPoint = hitPoint;
+      dragTangentDir = linalg::normalize(linalg::cross(selectionDisc.normal, radiusVec));
+
+      dragNormalDir = linalg::normalize(context.eyeLocation - hitPoint);
+      dragTangentDir = linalg::normalize(dragTangentDir - linalg::dot(dragTangentDir, dragNormalDir) * dragNormalDir);
+
+      SPDLOG_DEBUG("Drag start point: {} {} {}", dragStartPoint.x, dragStartPoint.y, dragStartPoint.z);
+    }
+  }
+
+  // Called when handle released from IGizmoCallbacks
+  virtual void released(InputContext &context, Handle &handle) {
+    size_t index = getHandleIndex(handle);
+    SPDLOG_DEBUG("Handle {} ({}) released", index, getAxisDirection(index, dragStartTransform));
+  }
+
+  // Called every update while handle is being held from IGizmoCallbacks
+  virtual void move(InputContext &context, Handle &handle) {
+
+    auto handleIndex = getHandleIndex(handle);
+    auto &selectionDisc = handleSelectionDiscs[handleIndex];
+
+    float d;
+    if (intersectPlane(context.eyeLocation, context.rayDirection, selectionDisc.center, dragNormalDir, d)) {
+      float3 hitPoint = context.eyeLocation + d * context.rayDirection;
+      float3 deltaVec = dragStartPoint - hitPoint;
+
+      float delta = linalg::dot(deltaVec, dragTangentDir);
+      // Use this to check direction vector of tangent and delta for the effective distance moved along the tangent
+      // SPDLOG_DEBUG("dragTangentDir {} {} {}", dragTangentDir.x, dragTangentDir.y, dragTangentDir.z);
+      // SPDLOG_DEBUG("Delta: {}", delta);
+      float sinTheta = std::sin(delta);
+      float cosTheta = std::cos(delta);
+
+      float4x4 rotationMat;
+
+      switch (handleIndex) {
```

Comment:
You can use something like
```
float3 axis;
axis[handleIndex] = 1.0f;
linalg::rotation_matrix(linalg::rotation_quat(axis, delta));
```
to keep it a bit simpler/shorter (the extra calculations don't matter too much for gizmos here)